test/test_multidevice_pipeline.cpp

```diff
@@ -0,0 +1,185 @@
+  TensorView* tv4a = set(tv4);
+  TensorView* tv5 = sum(tv4a, {0});
+
+  TensorView* tv6 = set(tv2a);
+  TensorView* tv7 = sum(tv6, {0});
+  fusion->addOutput(tv7);
+
+  TensorView* tv8 = set(tv3);
+  TensorView* tv9 = set(tv5);
+  TensorView* tv10 = set(tv7);
+  TensorView* tv11 = add(tv8, tv9);
+  TensorView* tv12 = add(tv11, tv10);
+  TensorView* tv13 = sum(tv12, {0});
+  fusion->addOutput(tv13);
+
+  // ===========================================================
+  //        PIPELINE SCHEDULING
+  // ===========================================================
+  /* Each TensorView must be assigned to one and only one stage
+     WAR: if an intermediate TensorView is automatically added
+          in the Fusion during Fusion definition,
+          it also needs to be assigned manually to a stage */
+  PipelineStageDescriptor stage0_, stage1_, stage0, stage1, stage2, stage3,
+      stage4;
+  stage0_.addVal({tv0_, tv1_});
+  stage1_.addVal({tv2_, tv3_});
+  stage0.addVal({tv0, tv1});
+  stage1.addVal({tv2, tv2a, tv3});
+  stage2.addVal({tv4, tv4a, tv5});
+  stage3.addVal({tv6, tv7});
+  stage4.addVal({tv8, tv9, tv10, tv11, tv12, tv13});
+
+  // binding each stage to a device mesh
+  stage0_.mesh = {5};
+  stage1_.mesh = {2, 4};
+  stage0.mesh = {0};
+  stage1.mesh = {0, 1, 4};
+  stage2.mesh = {1, 3};
+  stage3.mesh = {2};
+  stage4.mesh = {4, 5};
+
+  PipelineDescriptor descriptor{.stage_descriptors{
+      std::move(stage0_),
+      std::move(stage1_),
+      std::move(stage0),
+      std::move(stage1),
+      std::move(stage2),
+      std::move(stage3),
+      std::move(stage4)}}; // the order doesn't matter
+
+  pipeline = std::make_unique<Pipeline>(fusion.get(), std::move(descriptor));
+
+  // Create input tensors.
+  // Note: each process is binded to a different GPU
+  // Note: the concrete values are only used at the relevant ranks
+  inputs = {
+      at::randn({3096, 1123}, tensor_options),
+      at::randn({2048, 73, 81}, tensor_options)};
+
+  validate();
+}
+
+DeviceMesh mesh0({0});
+DeviceMesh mesh1({1});
+DeviceMesh mesh2({0, 1, 2, 3});
+DeviceMesh mesh3({0, 2, 3});
+auto all_meshes = ::testing::Values(mesh0, mesh1, mesh2, mesh3);
+
+TEST_P(PipelineTestTwoStages, Communication) {}
+
+INSTANTIATE_TEST_SUITE_P(
+    Gather,
+    PipelineTestTwoStages,
+    ::testing::Combine(
+        all_meshes,
+        all_meshes,
+        ::testing::Values(true),
+        ::testing::Values(false)));
+
+INSTANTIATE_TEST_SUITE_P(
+    Scatter,
+    PipelineTestTwoStages,
+    ::testing::Combine(
+        all_meshes,
+        all_meshes,
+        ::testing::Values(false),
+        ::testing::Values(true)));
+
+INSTANTIATE_TEST_SUITE_P(
+    Bcast,
+    PipelineTestTwoStages,
+    ::testing::Combine(
+        all_meshes,
+        all_meshes,
+        ::testing::Values(false),
+        ::testing::Values(false)));
+
+INSTANTIATE_TEST_SUITE_P(
+    Bcast_sharded,
+    PipelineTestTwoStages,
+    ::testing::Combine(
+        ::testing::Values(mesh2),
```

Comment:
What is this testing? It looks like no communication is necessary since tv1 and tv2 have the same mesh and sharding.

Reply:
Well spotted thank you. It's a mistake, it needs to be two different meshes of same size. Let me fix it